frontend/src/pages/Task.tsx

```diff
@@ -0,0 +1,630 @@
+                    />
+                </div>
+            </div>
+        </Card>
+    );
+}
+
+function TaskValidation({ task }: { task: TaskInfo }) {
+    // State Management for Component
+    const isActive = task.output_grammar ? true : false;
+    const [validationActive, setValidationActive] = useState<boolean>(isActive);
+    const [editingGrammer, setEditingGrammar] = useState<boolean>(!isActive);
+    const [taskGrammar, setTaskGrammar] = useState<string>(
+        `${task.output_grammar?.grammar_user_code ? task.output_grammar?.grammar_user_code : ""}`
+    );
+
+    // Query and Mutation Hooks
+    const [updateGrammarModelAction] = useUpdateGrammarModeMutation();
+
+    // Event Handlers
+    const onDownload = () => {
+        if (!task.output_grammar?.grammar_generated) {
+            console.log("Error: No generated grammar was loaded from the server.");
+            return;
+        }
+        const file = new File([task.output_grammar?.grammar_generated], "grammar.gbnf", {
+            type: "text/plain",
+        });
+
+        const link = document.createElement("a");
+        const url = URL.createObjectURL(file);
+
+        link.href = url;
+        link.download = file.name;
+        document.body.appendChild(link);
+        link.click();
+        document.body.removeChild(link);
+        window.URL.revokeObjectURL(url);
+    };
+
+    const onGenerate = () => {
+        try {
+            const srcFile = ts.createSourceFile("source.ts", taskGrammar, ts.ScriptTarget.ESNext);
+            const ifaces: Array<string> = [];
+            srcFile.forEachChild((child) => {
+                if (ts.isInterfaceDeclaration(child)) {
+                    ifaces.push(child.name.getText(srcFile));
+                }
+            });
+            const grammarArray = compile(taskGrammar, ifaces[0]);
+            const grammarFile = serializeGrammar(grammarArray);
+
+            updateGrammarModelAction({
+                task: task.name,
+                grammar: {
+                    grammar_user_code: taskGrammar,
+                    grammar_generated: grammarFile,
+                },
+            });
+
+            setEditingGrammar(false);
+        } catch (e) {
+            console.log(`Failed to compile or serialize grammar definition: ${e}`);
+        }
+    };
+
+    return (
+        <Card>
+            <div className=" flex flex-row gap-2 items-center">
+                <div className="mr-auto">
+                    <h3 className="text-lg font-semibold leading-none">Output Validation</h3>
+                </div>
+                {!validationActive ? (
+                    <Button
+                        size="medium"
+                        style="minimal"
+                        buttonText="Disabled"
+                        color="default"
+                        outline={true}
+                        onAction={() => setValidationActive(!validationActive)}
```

Comment:
In general, b/c the setter returned from `useState` executes async, you should not depend on the getter returned from the `useState` as the parameter to the setter, or else you could be reading a stale value by the time the setter actually runs. This happens if you have some events on the page which run slowly or block the main thread.

You should instead set the value directly, or if you want to do a toggle action, you do something like

```
setValidationActive(currentValidation => !currentValidation);
```

In this case, I'd just use `setValidationActive(true)` here directly